host_source.go

```diff
@@ -472,12 +472,106 @@ func (h *HostInfo) ScyllaShardAwarePortTLS() uint16 {
 	return h.scyllaShardAwarePortTLS
 }
 
-// Polls system.peers at a specific interval to find new hosts
+type ReplicaInfo struct {
+	hostId  UUID
+	shardId int
+}
+
+type TabletInfo struct {
+	mu           sync.RWMutex
+	keyspaceName string
+	tableId      UUID
+	lastToken    int64
+	tableName    string
+	tabletCount  int
+	newReplicas  []ReplicaInfo
+	replicas     []ReplicaInfo
+	stage        string
+}
+
+func (t *TabletInfo) KeyspaceName() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.keyspaceName
+}
+
+func (t *TabletInfo) TableId() UUID {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.tableId
+}
+
+func (t *TabletInfo) LastToken() int64 {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.lastToken
+}
+
+func (t *TabletInfo) TableName() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.tableName
+}
+
+func (t *TabletInfo) TabletCount() int {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.tabletCount
+}
+
+func (t *TabletInfo) NewReplicas() []ReplicaInfo {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.newReplicas
+}
+
+func (t *TabletInfo) Replicas() []ReplicaInfo {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.replicas
+}
+
+func (t *TabletInfo) Stage() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.stage
+}
+
+// Search for place in tablets table with specific Keyspce and Table name
+func findTablets(tablets []*TabletInfo, k string, t string) int {
+	l := -1
+	for i, tablet := range tablets {
+		if tablet.KeyspaceName() == k && tablet.TableName() == t {
+			l = i
+			break
+		}
+	}
+
+	return l
+}
+
+// Search for place in tablets table for token starting from index l
+func findTabletForToken(tablets []*TabletInfo, token token, l int) *TabletInfo {
+	r := l + tablets[l].TabletCount() - 1
+	for l < r {
+		m := (l + r) / 2
```

Comment:
Won't this `l+r` overflow is `l`, `r` are big?

Reply:
It could happen, I changed it a bit to prevent that

Reply:
Isn't `m := l + (r-l)/2` still incorrect? For:
```
l, r - signed 64-bit ints
r = 2^63 - 5
l = -(2^63-5)
r - l = 2^63 - 5 - (-(2^63-5))
r - l = 2^63 - 5 + 2^63 - 5
r - l = 2^64 - 10 > range of 64-bit signed int
```
(some other context: https://stackoverflow.com/questions/19106350/explanation-of-the-safe-average-of-two-numbers)

It's interesting that the tests didn't catch the problem, so either they are not sufficient enough to catch the bug or this code is actually correct for some reason.

Reply:
yeah it is incorrect it only solves the problem for two positive or two negative variables, I think that the tests are not sufficient enough to catch this bug